src/js/promise.js

```diff
@@ -0,0 +1,229 @@
+'use strict';
+
+function Promise(fn) {
+  if (!(this instanceof Promise))
+    throw new TypeError('undefined is not a promise');
+  if (typeof fn !== 'function')
+    throw new TypeError('Promise resolver undefined is not a function');
+  this._state = 0;
+  this._handled = false;
+  this._value = undefined;
+  this._deferreds = [];
+
+  doResolve(fn, this);
+}
+
+function handle(self, deferred) {
+  while (self._state === 3) {
```

Comment:
Use upper case global var instead of magic number 0, 1, 2, 3, or add comment to these numbers?